components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,409 @@
+ * last processed portion of data
+ *
+ * @param task_id - Id of the processing task
+ * @param chain_id - Id of the Chain, which the current data portion belongs to
+ * @param hex_account_address - Account address (whith padding), which the
+ * current data portion belongs to
+ * @param logs - chain_id and account address related Log data portion
+ * @param rawlogs - chain_id and account address related Log data portion
+ * @param error - error code
+ * @param error_message - error message
+ */
+void EthAllowanceManager::OnGetAllowances(
+    const int& task_id,
+    const std::string& chain_id,
+    const std::string& hex_account_address,
+    const std::vector<Log>& logs,
+    base::Value rawlogs,
+    mojom::ProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::ProviderError::kSuccess) {
+    if (tasks_.find(task_id) != tasks_.end()) {
+      tasks_[task_id]->MarkComplete();
+      MergeAllResultsAnCallBack();
+    }
+    return;
+  }
+
+  std::vector<Log> logs_tmp{logs};
+  std::sort(logs_tmp.begin(), logs_tmp.end(),
+            [](const Log& first, const Log& second) -> bool {
+              return std::tie(first.block_number, first.log_index) <
+                     std::tie(second.block_number, second.log_index);
+            });
+
+  // Collection of the latest allowances per contract & spender & approver.
+  std::map<std::string, std::tuple<uint256_t, mojom::AllowanceInfoPtr>>
+      allowance_map;
+  // Put cached data into the map first if available.
+  LoadCachedAllowances(chain_id, hex_account_address, allowance_map);
+  for (const auto& log_item : logs_tmp) {
+    // Skip pending logs.
+    if (log_item.block_number == 0) {
+      continue;
+    }
+
+    if (log_item.topics.size() < kTopicElementsCountToCheck) {
+      continue;
+    }
+    const auto current_map_key = GetAllowanceMapKey(
+        log_item.address, log_item.topics[1], log_item.topics[2]);
+
+    uint256_t parsed_amount{0};
+    if (!HexValueToUint256(log_item.data, &parsed_amount)) {
+      continue;
+    }
+
+    if (parsed_amount > 0) {
+      // Replace if same key exists by the fresh allowance data.
+      allowance_map.insert_or_assign(
+          current_map_key,
+          std::make_tuple(log_item.block_number,
+                          mojom::AllowanceInfo::New(
+                              chain_id, log_item.address, log_item.topics[1],
+                              log_item.topics[2], log_item.data)));
+    } else if (allowance_map.find(current_map_key) != allowance_map.end()) {
+      allowance_map.erase(current_map_key);
+    }
+  }
+  std::vector<mojom::AllowanceInfoPtr> allowances_found;
+  uint256_t max_block_number{0};
+  for (const auto& [allowance_map_key, allowance_tuple] : allowance_map) {
+    allowances_found.push_back(std::get<1>(allowance_tuple).Clone());
+    if (std::get<0>(allowance_tuple) > max_block_number) {
+      max_block_number = std::get<0>(allowance_tuple);
+    }
```

Comment:
We should not calculate max_block_number per logs. At least because last block is different per chain.
So in addition to https://github.com/brave/brave-core/pull/17807/files#r1209777404
- For each (address, chain) we should store last max block in prefs - last_block(address,chain)
- When DiscoverEthAllowancesOnAllSupportedChains starts we should for each chain_id get chain's current block - cur_block(chain)
- Then for each (address, chain) we should request logs with filter from last_block(address,chain)+1 (or earliest) to cur_block(chain).
- When logs are retrieved last_block(address,chain) is set to cur_block(chain).

Reply:
fixed